qa/tasks/ceph.py

```diff
@@ -1243,6 +1319,99 @@ def first_in_ceph_log(pattern, excludes):
             ),
         )
 
+@contextlib.contextmanager
+def create_osd_with_lvm(ctx, config):
+    '''
+    create osds with ceph-volume lvm
+    as of now doing only bluestore
+    TODO: filestore (because it needs extra journal device)
+    '''
+    remote_to_roles_to_devs = ctx.disk_config.remote_to_roles_to_dev
+    log.info(remote_to_roles_to_devs)
+    log.info('Using ceph-volume lvm for osd creation')
+    objstore = 'bluestore'
+    for section, keys in config['conf'].iteritems():
+        for key, val in keys.iteritems():
+            if key == 'osd objectstore':
+                if val == 'filestore':
+                    objstore = 'filestore'
+                    '''Filestore not implemented as of now'''
+                    log.error("ceph task with ceph-volume lvm not\
+			     implemented for filestore")
+                    return
+
+    log.info("Using objectstore {}".format(objstore))
+    ceph_lvm = "sudo ceph-volume lvm "
+    for remote, roles_to_dev in remote_to_roles_to_devs.iteritems():
+        for osd, dev in roles_to_dev.iteritems():
+            log.info("OSD={}, dev={}".format(osd, dev))
+            cmd = ceph_lvm + "create --" + objstore + \
+                " --data "+ dev
+            log.info("Running cmd={}".format(cmd))
+            remote.run(
+                args=cmd,
+            )
+            time.sleep(10)
+    try:
+        yield
+
+    finally:
+        teuthology.stop_daemons_of_type(ctx, "osd", config['cluster'])
+
+def _fix_roles_map(ctx, config):
+    if not hasattr(ctx, 'managers'):
+        ctx.managers = {}
+    ctx.daemons.use_systemd=True
+    ready_cluster = ctx.cluster
+    cluster_name = config['cluster']
+    if not hasattr(ctx, 'new_remote_role'):
+        new_remote_role = dict()
+        ctx.new_remote_role = new_remote_role
```

Comment:
why does this need to be preserved beyond this function?

Reply:
I think we should overwrite ctx.remote with new roles, ceph-volume will setup the osd roles which will not map to original roles defined in yaml by user( remote1: may have osd.4 , osd.1 depending on how they came up - just like ceph-ansible) , so for top layer tasks we have to fix that.